File: days/day_04/solutions/solution.ts
import * as path from 'path';
import * as fs from 'fs';

// Import utilities
const getInputPath = (day: number): string => "../data/input.txt";
const readInputRaw = (filePath: string): string => {
    const absPath = path.resolve(filePath);
    return fs.readFileSync(absPath, 'utf-8');
};

function countNeighbors(grid: string[], i: number, j: number, rows: number, cols: number): number {
    let count = 0;
    for (let di = -1; di <= 1; di++) {
        for (let dj = -1; dj <= 1; dj++) {
            if (di === 0 && dj === 0) continue;
            const ni = i + di;
            const nj = j + dj;
            if (ni >= 0 && ni < rows && nj >= 0 && nj < cols && grid[ni][nj] === '@') {
                count++;
            }
        }
    }
    return count;
}

function solve(inputData: string): [string, string] {
    const lines = inputData.trim().split('\n').filter(line => line.trim());
    const rows = lines.length;
    const cols = rows > 0 ? lines[0].length : 0;
    
    // Part 1: Count accessible rolls (fewer than 4 neighbors that are '@')
    let part1Count = 0;
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (lines[i][j] === '@') {
                const neighbors = countNeighbors(lines, i, j, rows, cols);
                if (neighbors < 4) {
                    part1Count++;
                }
            }
        }
    }
    
    // Part 2: Iteratively remove accessible rolls until none can be removed
    const grid: string[] = lines.map(line => line);
    
    let part2Count = 0;
    while (true) {
        const toRemove: [number, number][] = [];
        for (let i = 0; i < rows; i++) {
            for (let j = 0; j < cols; j++) {
                if (grid[i][j] === '@') {
                    const neighbors = countNeighbors(grid, i, j, rows, cols);
                    if (neighbors < 4) {
                        toRemove.push([i, j]);
                    }
                }
            }
        }
        
        if (toRemove.length === 0) {
            break;
        }
        
        // Remove all marked positions
        for (const [i, j] of toRemove) {
            grid[i] = grid[i].substring(0, j) + '.' + grid[i].substring(j + 1);
        }
        part2Count += toRemove.length;
    }
    
    return [part1Count.toString(), part2Count.toString()];
}

const data = readInputRaw("../data/input.txt");
const [part1, part2] = solve(data);
console.log(`Part 1: ${part1}`);
console.log(`Part 2: ${part2}`);

File: days/day_04/solutions/solution.py
from get_input import read_input_raw  # type: ignore
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve(
).parent.parent.parent.parent / 'utilities' / 'python'))


def count_neighbors(grid: list[str], i: int, j: int, rows: int, cols: int) -> int:
    """Count neighbors that are '@'."""
    neighbor_count = 0
    for di in [-1, 0, 1]:
        for dj in [-1, 0, 1]:
            if di == 0 and dj == 0:
                continue
            ni, nj = i + di, j + dj
            if 0 <= ni < rows and 0 <= nj < cols and grid[ni][nj] == '@':
                neighbor_count += 1
    return neighbor_count


def solve(input_data: str) -> tuple[str, str]:
    """Solve Day 04 parts 1 and 2."""
    lines = [line for line in input_data.strip().split('\n') if line.strip()]
    rows = len(lines)
    cols = len(lines[0]) if rows > 0 else 0

    # Part 1: Count accessible rolls (fewer than 4 neighbors that are '@')
    part1_count = 0
    for i in range(rows):
        for j in range(cols):
            if lines[i][j] == '@':
                neighbors = count_neighbors(lines, i, j, rows, cols)
                if neighbors < 4:
                    part1_count += 1

    # Part 2: Iteratively remove accessible rolls until none can be removed
    # Create mutable copy
    grid = [list(line) for line in lines]
    part2_count = 0

    while True:
        to_remove: list[tuple[int, int]] = []
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '@':
                    neighbors = count_neighbors(
                        [''.join(row) for row in grid], i, j, rows, cols)
                    if neighbors < 4:
                        to_remove.append((i, j))

        if not to_remove:
            break

        # Remove all marked positions
        for i, j in to_remove:
            grid[i][j] = '.'
        part2_count += len(to_remove)

    return str(part1_count), str(part2_count)


if __name__ == "__main__":
    data: str = read_input_raw("../data/input.txt")  # type: ignore
    part1, part2 = solve(data)  # type: ignore
    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")
